src/clp_ffi_py/ir/Query.hpp

```diff
@@ -0,0 +1,200 @@
+#ifndef CLP_FFI_PY_QUERY_HPP
+#define CLP_FFI_PY_QUERY_HPP
+
+#include <limits>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <clp/components/core/src/ErrorCode.hpp>
+#include <clp/components/core/src/ffi/encoding_methods.hpp>
+
+#include <clp_ffi_py/ExceptionFFI.hpp>
+#include <clp_ffi_py/ir/LogEvent.hpp>
+
+namespace clp_ffi_py::ir {
+/**
+ * This class defines a wildcard query, which includes a wildcard string and a
+ * boolean value to indicate if the match is case-sensitive.
+ */
+class WildcardQuery {
+public:
+    /**
+     * Initializes the wildcard query.
+     * @param wildcard Wildcard string.
+     * @param case_sensitive Case sensitive indicator.
+     */
+    WildcardQuery(std::string wildcard, bool case_sensitive)
+            : m_wildcard(std::move(wildcard)),
+              m_case_sensitive(case_sensitive){};
+
+    [[nodiscard]] auto get_wildcard() const -> std::string_view { return m_wildcard; }
+
+    [[nodiscard]] auto is_case_sensitive() const -> bool { return m_case_sensitive; }
+
+private:
+    std::string m_wildcard;
+    bool m_case_sensitive;
+};
+
+/**
+ * This class represents a search query, utilized for filtering log events in a
+ * CLP IR stream. The query could include a list of wildcard searches aimed at
+ * identifying certain log messages, and a timestamp range with a lower and
+ * upper bound. This class provides an interface to set up a search query, as
+ * well as methods to validate whether the query can be matched by a log event.
+ * Note that an empty wildcard query list will match any log within the range.
+ */
+class Query {
+public:
+    static constexpr ffi::epoch_time_ms_t const cTimestampMin{0};
+    static constexpr ffi::epoch_time_ms_t const cTimestampMax{
+            std::numeric_limits<ffi::epoch_time_ms_t>::max()};
+
+    /**
+     * Ideally, when searching an IR stream with a query, the search terminates
+     * once the timestamp exceeds the search upper bound. However, the timestamp
+     * might not be monotonically increasing in a CLP IR stream. It can be
+     * locally disordered due to thread contention. To safely stop searching, we
+     * need to ensure that the current timestamp has exceeded the search's upper
+     * bound by a reasonable margin. This margin defaults to the following
+     * constant. During the query initialization, it will be applied to the
+     * upper bound timestamp to generate a timestamp which indicates it is safe
+     * to terminate the search.
+     */
```

Comment:
```suggestion
    /**
     * Ideally, when searching an IR stream, the search would terminate once the
     * current log event's timestamp exceeds the query's time range. However,
     * timestamps in the IR stream might not be monotonically increasing; they
     * can be locally disordered due to thread contention. So to safely stop the
     * search, we need to ensure the current timestamp in the IR stream exceeds
     * the query's time range by a reasonable margin. This margin defaults to
     * the following constant. During the query initialization, it will be
     * applied to the upper bound timestamp to generate a timestamp which
     * indicates it is safe to terminate the search.
     */
```

Reply:
I think the description of the problem should go in the class' docstring, since it's relevant in multiple methods, not just in the specification of this constant.